src/src/PluginStructs/P104_data_struct.cpp

```diff
@@ -0,0 +1,2312 @@
+# include "../ESPEasyCore/ESPEasyRules.h"
+# include "../Globals/ESPEasy_time.h"
+
+# include <vector>
+# include <MD_Parola.h>
+# include <MD_MAX72xx.h>
+
+// Needed also here for PlatformIO's library finder as the .h file
+// is in a directory which is excluded in the src_filter
+
+# if defined(P104_USE_NUMERIC_DOUBLEHEIGHT_FONT) || defined(P104_USE_FULL_DOUBLEHEIGHT_FONT)
+void createHString(String& string); // Forward definition
+# endif // if defined(P104_USE_NUMERIC_DOUBLEHEIGHT_FONT) || defined(P104_USE_FULL_DOUBLEHEIGHT_FONT)
+void reverseStr(String& str);       // Forward definition
+
+/****************************************************************
+ * Constructor
+ ***************************************************************/
+P104_data_struct::P104_data_struct(MD_MAX72XX::moduleType_t _mod,
+                                   taskIndex_t              _taskIndex,
+                                   int8_t                   _cs_pin,
+                                   uint8_t                  _modulesize)
+  : mod(_mod), taskIndex(_taskIndex), cs_pin(_cs_pin), modules(_modulesize) {
+  if (Settings.InitSPI < 1) {
+    addLog(LOG_LEVEL_ERROR, F("DOTMATRIX: Required SPI not enabled. Initialization aborted!"));
+  } else {
+    P = new MD_Parola(mod, cs_pin, modules);
+  }
+}
+
+/*******************************
+ * Initializer/starter
+ ******************************/
+bool P104_data_struct::begin() {
+  if (!initialized) {
+    loadSettings();
+    initialized = true;
+  }
+
+  if ((P != nullptr) && (cs_pin > -1)) {
+    addLog(LOG_LEVEL_INFO, F("dotmatrix: P->begin() called"));
+    P->begin(expectedZones);
+    # ifdef P104_USE_BAR_GRAPH
+    pM = P->getGraphicObject();
+    # endif // ifdef P104_USE_BAR_GRAPH
+    return true;
+  }
+  return false;
+}
+
+# define P104_ZONE_SEP   '\x02'
+# define P104_FIELD_SEP  '\x01'
+# define P104_ZONE_DISP  ';'
+# define P104_FIELD_DISP ','
+
+/**************************************
+ * loadSettings
+ *************************************/
+void P104_data_struct::loadSettings() {
+  if (taskIndex < TASKS_MAX) {
+    LoadCustomTaskSettings(taskIndex, (uint8_t *)&StoredSettings, sizeof(StoredSettings.bufferSize));
+    uint16_t structDataSize = StoredSettings.bufferSize + sizeof(StoredSettings.bufferSize);
+    # ifdef P104_DEBUG_DEV
+    String log;
+
+    if (loglevelActiveFor(LOG_LEVEL_INFO) &&
+        log.reserve(54)) {
+      log  = F("P104: loadSettings stored Size: ");
+      log += structDataSize;
+      log += F(" taskindex: ");
+      log += taskIndex;
+      addLog(LOG_LEVEL_INFO, log);
+    }
+    # endif // ifdef P104_DEBUG_DEV
+    LoadCustomTaskSettings(taskIndex, (uint8_t *)&StoredSettings, structDataSize); // Read only actual data
+    StoredSettings.buffer[StoredSettings.bufferSize + 1] = '\0';                // Terminate string
+
+    uint8_t zoneIndex = 0;
+
+    {
+      String buffer;
+      buffer.reserve(StoredSettings.bufferSize + 1);
```

Comment:
Here you're creating a String object which essentially holds the buffer.
So you're using the memory twice, since both are allocated.
Do you really need to keep the `StoredSettings` in memory? And if so, isn't it easier to only store it as a String (thus heap allocated for the size it needs) ?
Then you can also forget the previous remark about using `get` and `operator[]` functions for accessing the data.
You only need to have that object present when loading/saving it.
For this you can also use the `SaveStringArray` function in ESPEasy_Storage.